First drop

jordigilh opened the pull request:
Missing README.md and a help message for the script.

jordigilh committed:
First drop: extends the boot partition and ext4 fs by shrinking the consecutive partition/lvm by a given amount

ygalblum reviewed (commented):
LGTM with one nit, the functions `is_device_mounted` and `is_ext4` are actually `enforce_...`  because they exit in case of false

jordigilh committed:
Fixes from testing the script on a Fedora 38 with an LVM partition

Signed-off-by: Jordi Gil <jgil@redhat.com>

jordigilh commented:
@ygalblum made changes after I noticed that the script was not achieving its goals on a Fedora 38. I will test it on a RHEL7 with an updated `sfdisk` version. I reached out to the `sfdisk` maintainer and he informed me that it's fine to run the latest version of `sfdisk` on RHEL7, no dependencies there. In fact, he recommended to run the latest `sfdisk` because he noticed issues in `--move-data` in previous versions that have been addressed in the latest versions of `sfdisk`.

jordigilh committed:
Renamed is_ext4 and is_device_mounted

Signed-off-by: Jordi Gil <jgil@redhat.com>

jordigilh committed:
Updated README.md

Signed-off-by: Jordi Gil <jgil@redhat.com>

ygalblum commented on extend.sh:
```diff
@@ -8,9 +8,24 @@ DEVICE_NAME=
+    echo "Script to extend the ext4 boot partition using BIOS by reducing the partition size of the adjacent partition by a given amount."
+    echo "It only works with ext4 file systems and supports both standard partitions with ext4 and partitions with LVM hosting ext4 file systems."
+    echo ""
+    echo "The script determines which partition number is the boot partition by looking for the boot flag. This process won't work with an EFI boot partition because the boot partition that contains the kernel and the initramfs are in a partition that does not have the flag."
```
Please break this line into two (you can split between the sentences)

ygalblum commented on extend.sh:
```diff
@@ -182,6 +219,24 @@ function resolve_device_name(){
 
+
+function deactivate_volume_group(){
+    volume_group_name=$(pvs "$DEVICE_NAME""$SUCCESSIVE_PARTITION_NUMBER" -o vg_name --noheadings|/usr/bin/sed 's/^[[:space:]]*//g')
```
Should these variables be local as well?

jordigilh replied:
Yeah, I prefer to keep them local to avoid conflicts.
Edit: Only when it makes sense though, not all `local` variables should be local I think, only those whose name could be reused elsewhere in another function. But in this case it applies since I'm using this name in another function.

jordigilh replied:
Done

ygalblum commented on extend.sh:
```diff
@@ -245,11 +303,12 @@ function shrink_physical_volume() {
-    local device=$DEVICE_NAME$SUCCESSIVE_PARTITION_NUMBER
-    current_partition_size_in_bytes=$(/usr/sbin/parted -m "$DEVICE_NAME" unit b print| /usr/bin/awk '/^'"$SUCCESSIVE_PARTITION_NUMBER"':/ {split($0,value,":"); print value[3]}'| /usr/bin/sed -e's/B//g')
+    local device=$DEVICE_NAME$1
+    current_partition_size_in_bytes=$(/usr/sbin/parted -m "$DEVICE_NAME" unit b print| /usr/bin/awk '/^'"$1"':/ {split($0,value,":"); print value[3]}'| /usr/bin/sed -e's/B//g')
```
Consider assigning `$1` to a local variable to make the code easier to read

jordigilh replied:
Done

ygalblum commented on extend.sh:
```diff
@@ -271,9 +330,18 @@ function resize_fs(){
-    local device=$DEVICE_NAME$SUCCESSIVE_PARTITION_NUMBER
-    device_type=$(get_device_type "$device")
+function shrink_partition() {
+    new_end_partition_size_in_bytes=$(calculate_new_end_partition_size_in_bytes "$1")
```
Consider assigning `$1` to a local variable to make the code easier to read

jordigilh replied:
Done

ygalblum reviewed (commented):
Minor nits

jordigilh committed:
Extended documentation to include information on how to build sfdisk from the source

Signed-off-by: Jordi Gil <jgil@redhat.com>

jordigilh committed:
Changes due to compatibility issues in RHEL7

Signed-off-by: Jordi Gil <jgil@redhat.com>

jordigilh committed:
Added a cleanup step at the end to enable the VG and fsck the boot file system

Signed-off-by: Jordi Gil <jgil@redhat.com>

ygalblum commented on extend.sh:
```diff
@@ -331,17 +331,19 @@ function resize_fs(){
-function shrink_target_partition(){
-    device_type=$(get_device_type "$DEVICE_NAME$SUCCESSIVE_PARTITION_NUMBER")
+function shrink_adjacent_partition(){
+    local device_type=
```
I think the `=` sign should be removed here, right?

jordigilh replied:
Shellcheck didn't complain, but I can remove it.

ygalblum commented on extend.sh:
```diff
@@ -422,14 +424,43 @@ function extend_boot_partition() {
     fi
 }
 
+function activate_volume_group(){
```
`pvs` and `vgchange` - use the full path

jordigilh replied:
Done in both instances (`deactivate` and `activate volume group` functions)

ygalblum commented:
I don't have permissions to resolve the previous comments, so if you want to clear the view, you'll need to do it :)

jordigilh committed:
Added full path for pvs and vgchange commands

Signed-off-by: Jordi Gil <jgil@redhat.com>

ygalblum commented:
LGTM

swapdisk commented on README.md:
````diff
@@ -1,2 +1,66 @@
+
+* Replace your existing `sfdisk` with the new binary:
+```
+[root@localhost ~]# mv sfdisk.static /usr/bin/sfdisk
````
Using `mv` will result in the wrong SELinux file context. Use `cp` or `mv -Z` instead.

jordigilh replied:
Done

swapdisk commented on README.md:
```diff
@@ -1,2 +1,66 @@
+
+
+# Dependencies
+This script requires the following binaries to work:
```
Add to list...
* /usr/bin/wc

jordigilh replied:
Done

jordigilh committed:
Added /usr/bin/wc to list of binary dependencies and changed 'cp' instead of 'mv' to overwrite the sfdisk binary

Signed-off-by: Jordi Gil <jgil@redhat.com>

swapdisk commented on extend.sh:
```diff
@@ -0,0 +1,466 @@
+    ret=$(/usr/sbin/lvreduce --resizefs -L -"$INCREMENT_BOOT_PARTITION_SIZE" "$LOGICAL_VOLUME_DEVICE_NAME" 2>&1)
+    local status=$?
+    if [[ $status -ne 0 ]]; then
+        echo "Failed to shrink logical volume $LOGICAL_VOLUME_DEVICE_NAME: $err"
```
```suggestion
        echo "Failed to shrink logical volume $LOGICAL_VOLUME_DEVICE_NAME: $ret"
```
Same thing in a few other places in the script.

jordigilh committed:
Ensure captured output variable names match when referencing them as part of the printed error message

jordigilh committed:
Refactored code so that it does not shrink the partition, instead it checks if there is enough space after the adjacent partition, shifts the adjacent partition and finally extends the boot partition and the file system in it, either ext4 or xfs

Signed-off-by: Jordi Gil <jgil@redhat.com>

jordigilh commented:
@swapdisk I've updated the code to expect free disk space right after the adjacent partition to the boot partition, something like this:
```
parted /dev/vdb p free
Model: Virtio Block Device (virtblk)
Disk /dev/vdb: 32.2GB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags:

Number  Start   End     Size    Type      File system  Flags
        32.3kB  1049kB  1016kB            Free Space
 1      1049kB  3625MB  3624MB  primary   ext4         boot
 2      3625MB  14.4GB  10.7GB  extended
 5      3626MB  14.4GB  10.7GB  logical
        14.4GB  32.2GB  17.9GB            Free Space
```

I removed all code relative to shrinking the LV, PV and partition.

jordigilh committed:
Added shrink PV capability and determine shrinking size taking into account the free available space after the adjacent partition. If there's enough free space, there will not be any shrinking action taken

Signed-off-by: Jordi Gil <jgil@redhat.com>

jordigilh committed:
Updated readme.md and added diagram for better understanding

jordigilh committed:
Fixes from running inside an initramfs and during pre-mount

jordigilh committed:
Rename script to bigboot and minor changes in name in the script

ygalblum commented on extend.sh:
```diff
@@ -0,0 +1,510 @@
+
+get_device_type(){
+    local device=$1
+    val=$(lsblk "$device" -o type --noheadings 2>&1)
```
Please use full path

jordigilh replied:
Done

ygalblum commented on extend.sh:
```diff
@@ -0,0 +1,510 @@
+    if [[ $device_type == "lvm" ]]; then
+        # It's an LVM block device 
+        # Capture the LV device names. Since we'll have to shift the partition, we need to make sure all LVs are not mounted in the adjacent partition.
+        devices_to_check=$(/usr/sbin/lvm pvdisplay "$device" -m |grep "Logical volume" |awk '{print $3}')
```
Please use full path for `awk`

jordigilh replied:
and grep 😄

jordigilh replied:
Done

ygalblum commented on extend.sh:
```diff
@@ -0,0 +1,510 @@
+
+get_fs_type(){
+    local device=$1
+    ret=$(blkid "$device" -o udev | sed -n -e 's/ID_FS_TYPE=//p' 2>&1)
```
Please use full path

jordigilh replied:
Done

ygalblum commented on extend.sh:
```diff
@@ -0,0 +1,510 @@
+    cleanup
+}
+
+main "$@"
```
Please add an empty line

jordigilh replied:
Done

ygalblum reviewed (commented):
Some minor full path nits. 
We should consider merging this PR even if it's not ready yet to allow smaller and easier to review PRs. It's not like someone will start using this script the minute you merge

jordigilh committed:
Added full path for additional commands

Signed-off-by: Jordi Gil <jgil@redhat.com>

jordigilh commented:
> Some minor full path nits. We should consider merging this PR even if it's not ready yet to allow smaller and easier to review PRs. It's not like someone will start using this script the minute you merge

Agreed, I will merge it as it is and will raise a new PR as required.
